backends/backends.c

```diff
@@ -425,32 +510,43 @@ void *backends_main(void *ptr) {
 
     // ------------------------------------------------------------------------
     // select the backend type
-    int adjust_type = backend_set_type(&default_port,&backend_response_checker,&backend_request_formatter,type);
-    if (adjust_type < 0) {
+    BACKEND_TYPE work_type = backends_select_type(type);
+    if (work_type == BACKEND_TYPE_UNKNOWN) {
         goto cleanup;
-    } else if (adjust_type) {
-        if (adjust_type == 1) {
+    }
+
+    void (*backend_fp[]) (int *,int (**brc)(BUFFER *),int (**bqf)(BUFFER *, const char *, RRDHOST *, const char *, RRDSET *, RRDDIM *, time_t, time_t, BACKEND_OPTIONS),const char *) = {
+            backend_set_graphite_variables
+            ,backend_set_opentsdb_telnet_variables
+            ,backend_set_opentsdb_http_variables
+            ,backend_set_json_variables
+            ,backend_set_prometheus_variables
+            ,backend_set_kinesis_variables
+    };
+    backend_fp[work_type](&default_port,&backend_response_checker,&backend_request_formatter,type);
```

Comment:
This is very error prone to rely on index and order. Just make it a function - a function that returns appropriate `backend_set_` function based on `type`

Reply:
I am sorry, I did not understand the  "error prone" cited, this is a function pointer and I am calling a function based in the index that I received from the function type. We won't have error, because when the software reaches this point, it will have only valid values to call the function. This is not much different of what we have in the file netdata/web/api/web_api_v1.c, the unique difference is the fact that I did not the struct api_commands to call, I called it directly.

Reply:
Correct me if i am wrong - if you use `enum` you should **never** rely on `enum` value and on `enum` order.
 
That is what you are doing here.

If you change the order or value in the `enum` your code doesn't work. If you change the order in your array of funcs your code doesnt work again.

Reply:
I could understand better now your point now, so let me explain better what I did. I am using the enum as a kind of variable that limit the possible value that is accepted, where the first value means error and the other values are valid index, so there is not any problem to use the value instead the name, because they are valid index. I understand that in object orientation languages this is not normal. 
I am bringing with this new fix a possible solution that helps to avoid developer errors, because the developer will need to change more places now, but as you can see I am also using index number instead names. Case we do not agree with this solution, we will be forced to use names instead number, but this will create a function, that is not necessary, but it will be more readable.

Reply:
> but this will create a function, that is not necessary

i think it is, but i am not insist on it, dont create it ok, but please change the implementation - use `if/else` or (better) `switch` to set the setter func instead of creating array and using enum as index.

Reply:
i think connecting array order with enam actual values is not good, lets ask @mfundul @vlvkobal. Maybe it is false alarm from me

Reply:
No it was not, I understood your point, because developers can write wrong things and your point of view helped me to write a structure that at least force the user to think more one time before to invert anything.
Firstly I will explain the computer side about "function is not necessary", when I said this, I was considering that I already have everything to call the function, so to create another step is not necessary, on the other hand, when we see the human side, I agree that it would be more clear to have a switch with all the options. No less important,  I agree with you let us bring @mfundul and @vlvkobal to explain their opinion, everybody will grow with this and the software will be optimized, with more resources and of course it will be better!